### ISSUE TITLE
Bounds check removed from slice but not Vec

### ISSUE THREAD

**Issue by khuey** [rust-lang/rust#46227]
When optimizing rustc will sometimes fail to remove unnecessary bounds checks from accesses to a Vec, when it's capable of removing the bounds check if operating on an equivalent slice.

```rust
#[inline(never)]
fn print_elements(v: &Vec<u64>) {
    // Uncomment this to remove the bounds check.
    //let v = &v[..];
    let len = v.len();
    let mut i = 0;
    while len > i {
        println!("{}", v[i]);
        i += 1;
    }
}

fn main() {
    let v = vec![1,2,3,4];
    print_elements(&v);
}
```

This produces the following asm for the loop body (on Release Nightly on play.rust-lang.org)

```asm
playground::print_elements:
<snip prologue>
.LBB1_2:
        cmpq	%rbx, %r14
	jbe	.LBB1_5
	movq	%r12, 8(%rsp)
	movq	core::fmt::num::<impl core::fmt::Display for u64>::fmt@GOTPCREL(%rip), %rax
	movq	%rax, 16(%rsp)
	leaq	ref.4(%rip), %rax
	movq	%rax, 24(%rsp)
	movq	$2, 32(%rsp)
	movq	%r13, 40(%rsp)
	movq	$1, 48(%rsp)
	movq	%rbp, 56(%rsp)
	movq	$1, 64(%rsp)
	movq	%r15, %rdi
	callq	std::io::stdio::_print@PLT
	incq	%rbx
	addq	$8, %r12
	cmpq	%rbx, %r14
	ja	.LBB1_2
<snip epilogue>
	retq
.LBB1_5:
	leaq	panic_bounds_check_loc.1(%rip), %rdi
	movq	%rbx, %rsi
	movq	%r14, %rdx
	callq	core::panicking::panic_bounds_check@PLT
	ud2
```

I'm not very familiar with the compiler internals, but this seems like an LLVM deficiency.  A reasonably sophisticated jump threading pass ought to catch this.

**Comment by khuey** [comment]
(Though, to be clear, not doing the jump threading optimization here doesn't in and of itself explain why the bounds check is removed on slices)

**Comment by est31** [comment]
I've used similar code that creates the same behaviour:

```rust
#![crate_type="lib"]

pub fn map_on_the_elems(v: &Vec<u64>, b: fn(u64)) {
    // Uncomment this to remove the bounds check.
    //let v = &v[..];
    for i in 0 .. v.len() {
        b(v[i]);
    }
}
```

LLVM IR with vec:

```
; playground::map_on_the_elems
; Function Attrs: uwtable
define void @_ZN10playground16map_on_the_elems17h382723b074a63bf8E(%"alloc::vec::Vec<u64>"* noalias nocapture readonly dereferenceable(24), void (i64)* nocapture) unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
start:
  %.idx = getelementptr %"alloc::vec::Vec<u64>", %"alloc::vec::Vec<u64>"* %0, i64 0, i32 2
  %.idx.val = load i64, i64* %.idx, align 8
  %2 = icmp eq i64 %.idx.val, 0
  br i1 %2, label %bb5, label %bb2.i.lr.ph

bb2.i.lr.ph:                                      ; preds = %start
  %.idx4 = getelementptr %"alloc::vec::Vec<u64>", %"alloc::vec::Vec<u64>"* %0, i64 0, i32 0, i32 0, i32 0, i32 0
  %.idx4.val = load i64*, i64** %.idx4, align 8
  br label %bb6

bb5.loopexit:                                     ; preds = %"_ZN81_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..Index$LT$usize$GT$$GT$5index17h7a894b1ad960f72cE.exit"
  br label %bb5

bb5:                                              ; preds = %bb5.loopexit, %start
  ret void

bb6:                                              ; preds = %"_ZN81_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..Index$LT$usize$GT$$GT$5index17h7a894b1ad960f72cE.exit", %bb2.i.lr.ph
  %iter.sroa.0.014 = phi i64 [ 0, %bb2.i.lr.ph ], [ %3, %"_ZN81_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..Index$LT$usize$GT$$GT$5index17h7a894b1ad960f72cE.exit" ]
  %3 = add nuw i64 %iter.sroa.0.014, 1
  %4 = icmp ugt i64 %.idx.val, %iter.sroa.0.014
  br i1 %4, label %"_ZN81_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..Index$LT$usize$GT$$GT$5index17h7a894b1ad960f72cE.exit", label %panic.i, !prof !0

panic.i:                                          ; preds = %bb6
; call core::panicking::panic_bounds_check
  tail call void @_ZN4core9panicking18panic_bounds_check17h00738207f5476c9aE({ %str_slice, [0 x i8], i32, [0 x i8], i32, [0 x i8] }* noalias readonly dereferenceable(24) bitcast ({ %str_slice, i32, i32 }* @panic_bounds_check_loc.1 to { %str_slice, [0 x i8], i32, [0 x i8], i32, [0 x i8] }*), i64 %iter.sroa.0.014, i64 %.idx.val)
  unreachable

"_ZN81_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..Index$LT$usize$GT$$GT$5index17h7a894b1ad960f72cE.exit": ; preds = %bb6
  %5 = getelementptr inbounds i64, i64* %.idx4.val, i64 %iter.sroa.0.014
  %6 = load i64, i64* %5, align 8
  tail call void %1(i64 %6)
  %7 = icmp ult i64 %3, %.idx.val
  br i1 %7, label %bb6, label %bb5.loopexit
}
```

LLVM IR with slice:

```
; playground::map_on_the_elems
; Function Attrs: uwtable
define void @_ZN10playground16map_on_the_elems17h382723b074a63bf8E(%"alloc::vec::Vec<u64>"* noalias nocapture readonly dereferenceable(24), void (i64)* nocapture) unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
start:
  %.idx = getelementptr %"alloc::vec::Vec<u64>", %"alloc::vec::Vec<u64>"* %0, i64 0, i32 0, i32 0, i32 0, i32 0
  %.idx.val = load i64*, i64** %.idx, align 8, !alias.scope !0
  %.idx3 = getelementptr %"alloc::vec::Vec<u64>", %"alloc::vec::Vec<u64>"* %0, i64 0, i32 2
  %.idx3.val = load i64, i64* %.idx3, align 8, !alias.scope !5
  %2 = icmp eq i64 %.idx3.val, 0
  br i1 %2, label %bb6, label %bb8.preheader

bb8.preheader:                                    ; preds = %start
  br label %bb8

bb6.loopexit:                                     ; preds = %bb8
  br label %bb6

bb6:                                              ; preds = %bb6.loopexit, %start
  ret void

bb8:                                              ; preds = %bb8.preheader, %bb8
  %iter.sroa.0.012 = phi i64 [ %3, %bb8 ], [ 0, %bb8.preheader ]
  %3 = add nuw i64 %iter.sroa.0.012, 1
  %4 = getelementptr inbounds i64, i64* %.idx.val, i64 %iter.sroa.0.012
  %5 = load i64, i64* %4, align 8
  tail call void %1(i64 %5)
  %6 = icmp ult i64 %3, %.idx3.val
  br i1 %6, label %bb8, label %bb6.loopexit
}
```

The LLVM IR seems to have one crucial difference: in the case of the vec, it emits a call to Vec::index, while in the case of the slice, it does the bounds check directly.

At its definition, `Vec::index` is marked as inline, and in fact due to monomorphisation the `Vec::index` function is being put into the same compilation unit. So there is no "LTO will fix this" excuse :).

**Comment by est31** [comment]
It has something to do with LLVM refusing to do inlining. This still keeps a bounds check (and doesn't inline), even though I am using a slice here:

```rust
#![crate_type="lib"]

pub fn map_on_the_elems(v: &Vec<u64>, b: fn(u64)) {
    let len = v.len();
    let mut i = 0;
    while len > i {
        b(index(&v[..], i));
        i += 1;
    }
}

#[inline(always)]
fn index(v: &[u64], idx: usize) -> u64 {
    v[idx]
}
```

**Comment by khuey** [comment]
If I run the `opt` from rustc's llvm build on the llvm bitcode that rustc produces the jump-threading pass does indeed take out the bounds check.

**Comment by est31** [comment]
clang is capable of removing the bounds check, so this is our bug: https://godbolt.org/g/yGxGeS

**Comment by khuey** [comment]
I misspoke above (now edited), tip-llvm's `opt` is capable of fixing up the IR that rustc generates, while rust-llvm's `opt` is not.

**Comment by andjo403** [comment]
think this is fixed now from opt-level=1 in rustc 1.36.0 the bound checks is removed for the examples 
from opt-level=2 the assembler is the same for the examples see https://godbolt.org/z/yyn-gt

**Comment by khuey** [comment]
This got fixed at some point.
